venom/passes/branch_opt/branchOptCorrectScript.sml

```diff
@@ -0,0 +1,252 @@
+Theorem step_inst_jnz_swapped:
+  !inst s cond_var lbl1 lbl2 v.
+    inst.inst_opcode = JNZ /\
+    inst.inst_operands = [Var cond_var; Label lbl2; Label lbl1] /\
+    lookup_var cond_var s = SOME v
+  ==>
+    step_inst inst s =
+      if v <> 0w then OK (jump_to lbl2 s)
+      else OK (jump_to lbl1 s)
+Proof
+  rw[step_inst_def, eval_operand_def]
+QED
+
+(* ==========================================================================
+   Combined ISZERO + JNZ Equivalence
+
+   This is the core correctness theorem for the transformation.
+
+   Original sequence:
+     %t = iszero %x
+     jnz %t, @A, @B
+
+   Transformed sequence:
+     nop
+     jnz %x, @B, @A
+
+   We show these produce equivalent final states (same jump target).
+   ========================================================================== *)
+
+(* The key equivalence: iszero+jnz vs direct jnz with swapped targets *)
+Theorem iszero_jnz_swap_equiv:
+  !x:bytes32 lbl_a lbl_b s.
+    (* Original: if (iszero x) <> 0 then A else B *)
+    (if bool_to_word (x = 0w) <> 0w then jump_to lbl_a s else jump_to lbl_b s) =
+    (* Transformed: if x <> 0 then B else A *)
+    (if x <> 0w then jump_to lbl_b s else jump_to lbl_a s)
+Proof
+  rpt gen_tac >>
+  simp[bool_to_word_nonzero] >>
+  Cases_on `x = 0w` >> simp[]
+QED
+
+(* ==========================================================================
+   NOP Semantics
+   ========================================================================== *)
+
+Theorem step_inst_nop:
+  !inst s.
+    inst.inst_opcode = NOP ==>
+    step_inst inst s = OK s
+Proof
+  rw[step_inst_def]
+QED
+
+(* ==========================================================================
+   Block Transformation Correctness
+
+   When the transformation applies (iszero+jnz pattern), the transformed
+   block produces the same execution result.
+   ========================================================================== *)
+
+(* Helper: The transformation pattern detection matches what we expect *)
+Theorem transform_block_pattern_match:
+  !bb iszero_inst jnz_inst out_var orig_cond.
+    LENGTH bb.bb_instructions >= 2 /\
+    EL (LENGTH bb.bb_instructions - 2) bb.bb_instructions = iszero_inst /\
+    EL (LENGTH bb.bb_instructions - 1) bb.bb_instructions = jnz_inst /\
+    is_iszero_inst iszero_inst /\
+    get_single_output iszero_inst = SOME out_var /\
+    iszero_inst.inst_operands = [orig_cond] /\
+    jnz_uses_var jnz_inst out_var
+  ==>
+    ?new_jnz rest.
+      (transform_block bb).bb_instructions =
+        rest ++ [iszero_inst with inst_opcode := NOP] ++ [new_jnz] /\
+      new_jnz = swap_jnz_operands (replace_jnz_cond jnz_inst orig_cond) /\
+      LENGTH rest = LENGTH bb.bb_instructions - 2
+Proof
+  rw[transform_block_def] >>
+  simp[] >>
+  qexists_tac `TAKE (LENGTH bb.bb_instructions - 2) bb.bb_instructions` >>
+  simp[LENGTH_TAKE]
+QED
+
+(* Main correctness theorem: transform preserves execution semantics *)
+Theorem transform_block_execution_equiv:
```

Comment:
I expect to see a top level correctness theorem that has a short statement and actually calls the compiler pass. This one looks like a lemma that has the pass inlined.